sdks/go/pkg/beam/core/runtime/graphx/translate.go

```diff
@@ -473,6 +479,164 @@ func (m *marshaller) addNode(n *graph.Node) string {
 	return m.makeNode(id, m.coders.Add(n.Coder), n)
 }
 
+// expandReshuffle translates resharding to a composite reshuffle
+// transform.
+//
+// With proper runner support, the SDK doesn't need to do anything.
+// However, we still need to provide a backup plan in terms of other
+// PTransforms in the event the runner doesn't have a native implementation.
+//
+// In particular, the "backup plan" needs to:
+//
+//  * Encode the windowed element, preserving timestamps.
+//  * Add random keys to the encoded windowed element []bytes
+//  * GroupByKey (in the global window).
+//  * Explode the resulting elements list.
+//  * Decode the windowed element []bytes.
+//
+// While a simple reshard can be written in user terms, (timestamps and windows
+// are accessible to user functions) there are some framework internal
+// optimizations that can be done if the framework is aware of the reshard, though
+// ideally this is handled on the runner side.
+//
+// User code is able to write reshards, but it's easier to access
+// the window coders framework side, which is critical for the reshard
+// to function with unbounded inputs.
+func (m *marshaller) expandReshuffle(edge NamedEdge) string {
+	id := edgeID(edge.Edge)
+	var kvCoderID, gbkCoderID string
+	{
+		kv := makeUnionCoder()
+		kvCoderID = m.coders.Add(kv)
+		gbkCoderID = m.coders.Add(coder.NewCoGBK(kv.Components))
+	}
+
+	var subtransforms []string
+
+	in := edge.Edge.Input[0]
+
+	origInput := m.addNode(in.From)
+	// We need to preserve the old windowing/triggering here
+	// for re-instatement after the GBK.
+	preservedWSId := m.pcollections[origInput].GetWindowingStrategyId()
+
+	// Get the windowing strategy from before:
+	postReify := fmt.Sprintf("%v_%v_reifyts", nodeID(in.From), id)
+	m.makeNode(postReify, kvCoderID, in.From)
+
+	// We need to replace postReify's windowing strategy with one appropriate
+	// for reshuffles.
+	{
+		wfn := window.NewGlobalWindows()
+		m.pcollections[postReify].WindowingStrategyId =
+			m.internWindowingStrategy(&pb.WindowingStrategy{
+				// Not segregated by time...
+				WindowFn: makeWindowFn(wfn),
+				// ...output after every element is received...
+				Trigger: &pb.Trigger{
+					// Should this be an Always trigger instead?
```

Comment:
In answer to this comment, yes it should be Always

Reply:
Thanks!